File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/CartService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.CreateTicketDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.Ticket;
import at.ac.tuwien.sepm.groupphase.backend.entity.TicketOrder;

import java.util.List;

public interface CartService {

    /**
     * Adds the given to the cart of the logged-in user.
     *
     * @param userId the ID of the logged-in user
     * @param tickets the tickets that should be added to the cart
     */
    void addTicketsToCart(Long userId, List<CreateTicketDto> tickets);

    /**
     * Gets the cart of the currently logged-in user.
     *
     * @param userId the ID of the logged-in user
     * @return the cart TicketOrder of the currently logged-in user
     */
    TicketOrder getCart(Long userId);

    /**
     * Removes a ticket (by its ID) from the cart of the logged-in user.
     *
     * @param userId the ID of the logged-in user
     * @param ticketId the ID of the ticked which should be removed
     */
    void removeTicket(Long userId, Long ticketId);

    /**
     * Gets all purchased tickets for upcoming events of the currently logged-in user.
     *
     * @param userId the ID of the logged-in user
     * @return the purchased TicketOrder of the currently logged-in user
     */
    List<Ticket> getPurchasedTickets(Long userId);
}

File: frontend/src/app/services/cart.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Globals} from '../global/globals';
import {CreateTicket} from '../dtos/create-ticket';
import {Observable} from 'rxjs';
import {Cart} from '../dtos/cart';
import {PagedTicket, Ticket} from '../dtos/ticket';

@Injectable({
  providedIn: 'root'
})
export class CartService {

  private readonly cartBaseUri: string = `${this.globals.backendUri}/cart`;

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Adds multiple tickets to the cart of the currently logged-in user.
   *
   * @param tickets the tickets to be added to the logged-in users cart
   */
  addTicketsToCart(tickets: CreateTicket[]): Observable<void> {
    console.log(`Add ${tickets.length} tickets to cart.`);
    return this.httpClient.post<void>(this.cartBaseUri, tickets);
  }

  /**
   * Get the cart of the currently logged-in user.
   */
  getCart(): Observable<Cart> {
    console.log(`Get cart of logged-in user.`);
    return this.httpClient.get<Cart>(this.cartBaseUri);
  }

  /**
   * Remove ticket from the cart of the currently logged-in user.
   *
   * @param ticketId ID of the ticket which should be removed
   */
  removeTicketFromCart(ticketId: number): Observable<void> {
    console.log(`Remove ticket ${ticketId} from cart.`);
    return this.httpClient.delete<void>(`${this.cartBaseUri}/tickets/${ticketId}`);
  }

  /**
   * Get upcoming purchased Tickets of the currently logged-in user.
   */
  getUpcomingPurchasedTickets(): Observable<Ticket[]> {
    return this.httpClient.get<Ticket[]>(this.cartBaseUri + '/purchased/upcoming');
  }

  /**
   * Get past purchased Tickets of the currently logged-in user.
   */
  getPastPurchasedTickets(page: number = 0, size: number = 6): Observable<PagedTicket> {
    return this.httpClient.get<PagedTicket>(this.cartBaseUri + '/purchased/past' + '?page=' + page + '&size=' + size);
  }
}
